docs/designs/custom_allreduce_training_loop_support.md

```diff
@@ -0,0 +1,237 @@
+# Support Customizing the Training Loop Using AllReduce
+
+This document describes the design for supporting customizing
+the training loop when users use AllReduce distribution strategy.
+
+## Motivation
+
+Now, users need to define the forward computation, loss function,
+optimizer and dataset function in ElasticDL. ElasticDL provides
+the training loop with those definitions. Now ElasticDL only support
+Keras API to define the forward computation. It maybe not flexible for
+users to define complex models in CV or NLP. Sometimes, users need to
+customize the training loop to control the model iteration.
```

Comment:
Could you please give an example of the demand for customizing the training loop?

Reply:
Given a pre-trained network, users might only want to optimize a new set of output weights.